packages/astro/src/content/runtime.ts

```diff
@@ -87,12 +88,36 @@ export function createGetCollection({
 				const data = rawEntry.filePath
 					? updateImageReferencesInData(rawEntry.data, rawEntry.filePath, imageAssetMap)
 					: rawEntry.data;
-				const entry = {
-					...rawEntry,
-					data,
-					collection,
-				};
-				if(hasFilter && !filter(entry)) {
+				let entry;
+				if (!rawEntry.isModule) {
+					entry = {
+						...rawEntry,
+						data,
+						collection,
+					};
+				} else {
+					// @ts-expect-error	virtual module
+					const { default: contentModules } = await import('astro:content-module-imports');
+					const module = contentModules.get(rawEntry.filePath);
+					const resolvedComponent = await module();
+
+					entry = {
+						...rawEntry,
+						collection,
+						id: rawEntry.id,
+						slug: rawEntry.id,
+						body: rawEntry.body,
+						data: rawEntry.data,
+						async render() {
+							return {
+								Content: resolvedComponent.Content,
+								headings: resolvedComponent.getHeadings?.() ?? [],
+								remarkPluginFrontmatter: resolvedComponent.frontmatter ?? {},
+							};
+						},
+					};
```

Comment:
I think there's arguments for and against making this more like existing MDX (so like this one) or more like new markdown (which puts headings and frontmatter inside `rendered.metadata`).